tests/ui/openid_connect_mock.rs

```diff
@@ -596,10 +753,53 @@ fn run_mock_openid_connect_server() {
                         if let Some(session) = login_sessions.get(refresh_token) {
                             let user = get_user_for_session(&session, known_users)?;
+                            log_warning(&format!("session: {:?}", &session.id));
+                            // Check the intentional failure responses we might want to
+                            // impose on users and return the apprioriate HTTP response.
+                            match user.on_refresh_token_failure {
+                                None => {},
+                                Some(RefreshTokenFailureMode::NoResponse) => {
+                                    log_warning(&format!("Not responding to request"));
+                                    thread::park();
+                                },
+                                Some(RefreshTokenFailureMode::SlowResponse(dur)) => {
+                                    log_warning(&format!("Responding slowly after {} seconds", &dur.as_secs()));
+                                    thread::sleep(dur);
+                                }
+                                Some(RefreshTokenFailureMode::Error500Response) => {
+                                    log_warning("Responding with a 500");
+                                    return request
+                                        .respond(
+                                            Response::empty(StatusCode(500))
+                                        )
+                                        .map_err(|err| err.into());
+                                }
+                                Some(RefreshTokenFailureMode::Error503Response) => {
+                                    log_warning("Responding with a 503");
+                                    return request
+                                        .respond(
+                                            Response::empty(StatusCode(503))
+                                        )
+                                        .map_err(|err| err.into());
+                                }
+                                Some(err_mode) => {
+                                    log_warning(&format!("(Intentionally) returning error '{}' for user '{}'", err_mode, &session.id));
+                                    return request
+                                        .respond(
+                                            Response::empty(StatusCode(400))
+                                                .with_header(
+                                                    Header::from_str("Content-Type: application/json").unwrap(),
+                                                )
+                                                .with_data(format!("{{\"error\":\"{}\"}}", err_mode).as_bytes(), None),
+                                        )
+                                        .map_err(|err| err.into());
+                                }
+                            }
```

Comment:
You could just write this as b"..." since it's a constant string?
Or use the serde_json json! macro to ensure you are producing correct JSON.

Reply:
5e96304 converts it to using json\_serde